### PR TITLE
Fix 4662

### PR DESCRIPTION
**Pull request by smartacus**
This is the code from PR 4662 rebased against master. I will fix any hound notifications or failing specs that show up.

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://trialguides-fix-4662.spree-spree-9ed724b77adec04c49c2.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://trialguides-fix-4662.spree-spree-9ed724b77adec04c49c2.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

**Comment by smartacus** [comment]
@JDutil The last two syntax issues are fixed but hound isn't picking them up for some reason.

File: core/lib/spree/core/search/variant.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+            fields.each do |field|
+              query[:g][grouping_count.to_s][:c][condition_count.to_s] = {
+                a: { '0' => { name: field } },
+                p: 'cont',
+                v: { '0' => { value: part } }
+              }
+              condition_count += 1
+            end
+            grouping_count += 1
+          end
```
Comment (BenMorganIO) [inline core/lib/spree/core/search/variant.rb]: Some of this should probably be moved off into its own private methods and code comments would be appreciated. :+1: on the specs.

File: core/lib/spree/core/search/product.rb [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+          @products = @products.page(@params[:page]).per(@params[:per_page])
+        end
+
+        protected
+
+        def add_eagerload_scopes(scope)
+          # TL;DR Switch from `preload` to `includes` as soon as Rails starts honoring
+          # `order` clauses on `has_many` associations when a `where` constraint
+          # affecting a joined table is present (see
+          # https://github.com/rails/rails/issues/6769).
```
Comment (BenMorganIO) [inline core/lib/spree/core/search/product.rb]: Can you check if this is still existent? There's been a lot of bug fixes in AR. https://github.com/rails/rails/commits/master/activerecord

**Comment by BenMorganIO** [comment]
@smartacus Can you add this bug fix to the Changelog?

**Comment by smartacus** [comment]
@BenMorganIO I refactored the variant search method but I'm honestly not sure if it's any easier to understand. What do you think? I can still add comments to it as well.

File: core/lib/spree/core/search/product.rb (hunk 2 above)
Comment (smartacus) [inline core/lib/spree/core/search/product.rb]: I was unable to reproduce the issue using AR 4.2.3. I'll double-check tomorrow morning and remove the preload after I'm sure it's fixed.

Comment (BenMorganIO) [inline core/lib/spree/core/search/product.rb]: Thanks! :heart:

File: api/app/controllers/spree/api/v1/variants_controller.rb [hunk 3]
```diff
@@ -20,12 +22,27 @@ def destroy
...
-        # The lazyloaded associations here are pretty much attached to which nodes
-        # we render on the view so we better update it any time a node is included
-        # or removed from the views.
         def index
-          @variants = scope.includes({ option_values: :option_type }, :product, :default_price, :images, { stock_items: :stock_location })
-            .ransack(params[:q]).result.page(params[:page]).per(params[:per_page])
+          # The lazyloaded associations here are pretty much attached to which nodes
+          # we render on the view so we better update it any time a node is included
+          # or removed from the views.
+          base_scope = scope.includes(
```
Comment (BenMorganIO) [inline api/app/controllers/spree/api/v1/variants_controller.rb]: @smartacus maybe remove `base_scope`?

File: core/lib/spree/core/search/product.rb (hunk 2 above)
Comment (smartacus) [inline core/lib/spree/core/search/product.rb]: I was able to reproduce the issue using AR 4.2.3 and edge.

File: api/app/controllers/spree/api/v1/variants_controller.rb (hunk 3 above)
Comment (smartacus) [inline api/app/controllers/spree/api/v1/variants_controller.rb]: It is used on line 40.

File: core/lib/spree/core/search/product.rb (hunk 2 above)
Comment (BenMorganIO) [inline core/lib/spree/core/search/product.rb]: kk, thanks.

File: core/lib/spree/core/search.rb [hunk 4]
```diff
@@ -0,0 +1,3 @@
+require 'spree/core/search/base'
+require 'spree/core/search/product'
+require 'spree/core/search/variant'
```
[the remark this answers is not in the document]
Comment (keonjeo) [reply core/lib/spree/core/search.rb]: 👍  Could you add a newline here according to the hound advice?

**Pull request closed without merging**